Sidebranch

Samet-YAZAR opened the pull request:
Project has completed, waiting for any constructive feedback.

Samet-YAZAR committed:
Add new project in sidebrach

Samet-YAZAR committed:
Add gitignore

Samet-YAZAR committed:
Remove ignored files

Samet-YAZAR committed:
Write PersonService class methods

Samet-YAZAR committed:
Complete Program class,
Finish Person Book App

Samet-YAZAR committed:
Refactor

AhmetKuris commented on PersonBook/PersonBook/PersonBook/Entity/Person.cs:
```diff
@@ -0,0 +1,25 @@
+        public Person( string firstName, string lastName)
+        {
+            FirstName = firstName;
+            LastName = lastName;
```
Remove unnecessary comments

Samet-YAZAR replied:
Done

AhmetKuris commented on PersonBook/PersonBook/PersonBook/Program.cs:
```diff
@@ -0,0 +1,78 @@
+﻿using PersonBook.Entity;
+using PersonBook.Service;
```
Should be under namespace

Samet-YAZAR replied:
Done

AhmetKuris commented on PersonBook/PersonBook/PersonBook/Program.cs:
```diff
@@ -0,0 +1,78 @@
+    {
+       //
+        static void Main(string[] args)
+        {
```
Unnecessary Comments

Samet-YAZAR replied:
Done

AhmetKuris commented on PersonBook/PersonBook/PersonBook/Service/PersonService.cs:
```diff
@@ -0,0 +1,48 @@
+﻿using PersonBook.Entity;
```
Should be under namespace

Samet-YAZAR replied:
Done

AhmetKuris commented on PersonBook/PersonBook/PersonBook/Service/PersonService.cs:
```diff
@@ -0,0 +1,48 @@
+        public PersonService()
+        {
+            _personList = new List<Person>();
+        }
+        public void AddPerson(Person person)
+        {
+            person.Id = _random.Next(1000, 9999);
+            _personList.Add(person);
+        }
+
```
Maybe you can check if Id already exits or not.

Samet-YAZAR replied:
Done

durnasaban commented on PersonBook/PersonBook/PersonBook/Entity/Person.cs:
```diff
@@ -0,0 +1,25 @@
+{
+    class Person 
+    {
+       // [Required]
```
The properties already have setter,. so we don't need to populate them from constructor. You can remove constructors!

Samet-YAZAR replied:
Done

durnasaban commented on PersonBook/PersonBook/PersonBook/Program.cs:
```diff
@@ -0,0 +1,78 @@
+       //
+        static void Main(string[] args)
+        {
+            PersonService pservice = new PersonService();
```
You should use 'var' instead of PersonService like below.
`var pservice = new PersonService();`
Also, we shouldn't enshorten the variable names. Variable name could be `personService`

Samet-YAZAR replied:
Done

durnasaban commented on PersonBook/PersonBook/PersonBook/Program.cs:
```diff
@@ -0,0 +1,78 @@
+            while (true)
+            {
+                Console.WriteLine("\n" + 
+               "~~~~~~Welcome to Person Book Application~~~~~\n " +
```
You can move it to a variable!

Samet-YAZAR replied:
Done

durnasaban commented on PersonBook/PersonBook/PersonBook/Program.cs:
```diff
@@ -0,0 +1,78 @@
+                    case  "1":
+
+                            pservice.ListPersons().ForEach(p => {
+                                Console.WriteLine("********************************\n");
```
You can add one more method to pservice as "ShowPerson" and move this string into it. 
I mean here `"Id : " + p.Id + "\n" +
                                    "First Name : " + p.FirstName + "\n" +
                                    "Last Name : " + p.LastName`

Samet-YAZAR replied:
Done

durnasaban commented on PersonBook/PersonBook/PersonBook/Program.cs:
```diff
@@ -0,0 +1,78 @@
+                        string name = Console.ReadLine();
+                        string surname = Console.ReadLine();
+                        Person newPerson =new Person(name,surname);
+                        pservice.AddPerson(newPerson);
```
AddPErson method can get name and surname values. You don't need to create new Person instance here

Samet-YAZAR replied:
Done

durnasaban commented on PersonBook/PersonBook/PersonBook/Program.cs:
```diff
@@ -0,0 +1,78 @@
+                    case "3":
+
+                        Console.WriteLine("Please enter an ID and new Name, Surname ");
+                        int inputId = int.Parse(Console.ReadLine());
```
Too dangerous!!!!!
You should use Int32.TryParse instead of int.parse because a user can enter non numeric characters.

durnasaban commented on PersonBook/PersonBook/PersonBook/Program.cs:
```diff
@@ -0,0 +1,78 @@
+                        string inputName = Console.ReadLine();
+                        string inputSurname = Console.ReadLine();
+                        Person pUpdate = new Person(inputId, inputName, inputSurname);
+                        pservice.UpdatePerson(pUpdate);
```
Update person should get id, name, surname!

durnasaban commented on PersonBook/PersonBook/PersonBook/Program.cs:
```diff
@@ -0,0 +1,78 @@
+
+                        Console.WriteLine("Please enter an ID and new Name, Surname ");
+                        int inputId = int.Parse(Console.ReadLine());
+                        string inputName = Console.ReadLine();
```
You can inform those the user this one is name and this one is surname.
 I mean that the user should see the name and surname during entering the values:
name: 
surname:

durnasaban commented on PersonBook/PersonBook/PersonBook/Program.cs:
```diff
@@ -0,0 +1,78 @@
+                    case "4":
+
+                        Console.WriteLine("Please enter an ID ");
+                        int deleteId = int.Parse(Console.ReadLine());
```
Here is dangerous too!!

durnasaban commented on PersonBook/PersonBook/PersonBook/Service/IPersonService.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    interface IPersonService<Person>
+    {
+        public List<Person> ListPersons();
```
You can use ICollection instead of List!

durnasaban replied:
You shouldn't use `public` in interfaces!!!

Samet-YAZAR replied:
Done

durnasaban commented on PersonBook/PersonBook/PersonBook/Service/PersonService.cs:
```diff
@@ -0,0 +1,48 @@
+    class PersonService : IPersonService<Person>
+    {
+        private List<Person> _personList;
+        private readonly Random _random = new Random();
```
You can create a method which can return new id and that method can calculate the id by adding 1 to the biggest id in the list.

Samet-YAZAR replied:
Done

durnasaban commented on PersonBook/PersonBook/PersonBook/Service/PersonService.cs:
```diff
@@ -0,0 +1,48 @@
+{
+    class PersonService : IPersonService<Person>
+    {
+        private List<Person> _personList;
```
`ICollection` instead of `List`

Samet-YAZAR replied:
Done

durnasaban commented on PersonBook/PersonBook/PersonBook/Service/PersonService.cs:
```diff
@@ -0,0 +1,48 @@
+
+        public void DeletePerson(int id)
+        {
+            Person p = _personList.FirstOrDefault(i => i.Id == id);
```
You should use `var` instead of `Person`!

`var person = _personList.FirstOrDefault(i => i.Id == id);`

Samet-YAZAR replied:
Done

durnasaban commented on PersonBook/PersonBook/PersonBook/Service/PersonService.cs:
```diff
@@ -0,0 +1,48 @@
+
+        public void UpdatePerson(Person person)
+        {
+            Person p = _personList.FirstOrDefault(i => i.Id == person.Id);
```
You should check person is null or empty. The user can input an id that couldn't be in the list!

Samet-YAZAR replied:
Done

Samet-YAZAR committed:
Restore methods except the random generation of a Person Id

Samet-YAZAR committed:
Add Id Generator Class,
Modify Exception Class in project

Samet-YAZAR committed:
Refactor effectiveness, remove useless lines

Samet-YAZAR committed:
Refactor

durnasaban commented on PersonBook/PersonBook/PersonBook/Entity/Person.cs:
```diff
@@ -1,25 +1,11 @@
 
 namespace PersonBook.Entity
 {
     class Person 
```
Marking entities with `public` would be a better approach.

Samet-YAZAR replied:
Done

durnasaban commented on PersonBook/PersonBook/PersonBook/Program.cs:
```diff
@@ -1,78 +1,88 @@
         {
-            PersonService pservice = new PersonService();
+            var personService = new PersonService();
+            bool flag = true;
```
`flag` as a name seems too common. That could be a more specific name. 
`isContinued` or `isInApp` etc..

Samet-YAZAR replied:
Done

durnasaban commented on PersonBook/PersonBook/PersonBook/Program.cs:
```diff
@@ -1,78 +1,88 @@
                 string option = Console.ReadLine();
-              
+
                 switch (option)
```
there shouldn't be any logic in switch case statement. 
```
switch(item){
  case 1:
    do();
   break;
case 2:
   doElse();
break;
}
```

Samet-YAZAR replied:
Done

durnasaban commented on PersonBook/PersonBook/PersonBook/Program.cs:
```diff
@@ -1,78 +1,88 @@
-                        pservice.DeletePerson(deleteId);
+                        Console.WriteLine("Id: ");
+                        int number;
+                        Int32.TryParse(Console.ReadLine(),out number);
```
you don't need to define `int number`. you can do it like below.
`Int32.TryParse(Console.ReadLine(),out var number)`

durnasaban commented on PersonBook/PersonBook/PersonBook/Service/PersonService.cs:
```diff
@@ -1,24 +1,33 @@
-    class PersonService : IPersonService<Person>
+    using Entity;
+
+     class PersonService : IPersonService<Person>
```
`IPersonService<Person>` doesn't need to be generic. it can be just `IPersonService`

Samet-YAZAR replied:
Done

durnasaban commented on PersonBook/PersonBook/PersonBook/Service/PersonService.cs:
```diff
@@ -31,17 +40,17 @@ public void DeletePerson(int id)
-            Person p = _personList.FirstOrDefault(i => i.Id == person.Id);
-            p.FirstName = person.FirstName;
-            p.LastName = person.LastName;
+            Person p = _personList.FirstOrDefault(i => i.Id == Id);
```
that should be `var person = _personList.Fir....`
And you should check whether the person is null or not!

Samet-YAZAR replied:
Done

Samet-YAZAR committed:
change flag to isContinued,
Move foreach from Main,
use out var number,
Use pablic marking with entity

Samet-YAZAR committed:
Update PersonService.cs

durnasaban commented on PersonBook/PersonBook/PersonBook/Program.cs:
```diff
@@ -0,0 +1,103 @@
+
+                switch (option)
+                {
+                    case "1":
```
You should wrap items in cases like below.
```
private const string Show = "1";
private const string Add = "2";
.....
case Show:
    .....
    break;
case Add:
   ....
   break;
````
```

Samet-YAZAR replied:
Done

durnasaban commented on PersonBook/PersonBook/PersonBook/Service/PersonService.cs:
```diff
@@ -0,0 +1,68 @@
+
+        public void AddPerson(string name, string surname)
+        {
+            int Id = IdGenerator();
```
variable names should start with lowercase.
`var id = IdGenerator();`

durnasaban replied:
Method names should start with verbs.
That should be `GenerateId()`

Samet-YAZAR replied:
Done

durnasaban commented on PersonBook/PersonBook/PersonBook/Service/PersonService.cs:
```diff
@@ -0,0 +1,68 @@
+        public void DeletePerson(int id)
+        {
+
+            var personToRemove = _personList.SingleOrDefault(p => p.Id == id);
```
SingleOrDefault is not a good solution here. You could user FirstOrDefault instead and then check if it's null. if it's null you can let the user know the issue and then return;

durnasaban replied:
variable name could be just `person`

Samet-YAZAR replied:
Done

Samet-YAZAR committed:
Change SingleOrDefaut method to firstOrDefault,
Move hard coded case numbers to variable,
Rename IdGenerator method to GenerateID

Samet-YAZAR committed:
Merge branch 'sidebranch' of https://github.com/Samet-YAZAR/PersonBook into sidebranch

Samet-YAZAR committed:
Refactoring

Samet-YAZAR committed:
Import Entity in IPersonService
Remove Generic Person extension from IPersonService
